lib/Sema/CSApply.cpp

```diff
@@ -2406,8 +2411,90 @@ namespace {
+      }
+
+      // CSGen has already ensured that the first argument is a string, so skip
+      // the first argument.
+      for (unsigned i = 1, e = args->getNumElements(); i != e; ++i) {
+        auto argExpr = args->getElement(i);
+        auto argType = cs.getType(argExpr);
+        auto argLoc = argExpr->getLoc();
+        auto argNameAndLowering = tf::GraphOperationInfo::decodeArgumentName(
+            args->getElementName(i).str());
+        if (!argNameAndLowering) {
+          diagnose(argLoc, "argument name has invalid modifier");
+          return nullptr;
+        }
+        auto argLowering = argNameAndLowering->second;
+        switch (argLowering) {
+        case tf::GraphOperationInfo::ArgumentLowering::Input:
+          // CSGen has already ensured that all inputs conform to
+          // TensorArrayProtocol.
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::NormalAttribute:
+          if (atc.classifyNormalAttribute(argType) ==
+              tf::AttributeTypeClassifier::Normal::Unsupported) {
+            diagnose(argLoc, "attribute requires Bool, Int64, Double, Float, "
+                             "array thereof, [TensorShape?], or Function, "
+                             "but got type '" + argType->getString() + "'");
+            return nullptr;
+          }
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::TensorAttribute:
+          // See the comment on the declaration of the enum case
+          // `GraphOperationInfo::ArgumentLowering::TensorAttribute` for more
+          // information on why these are not allowed in #tfop's.
+          diagnose(argLoc, "$tensor attributes are not allowed in #tfop's");
+          return nullptr;
+        case tf::GraphOperationInfo::ArgumentLowering::ShapeAttribute:
+          if (atc.classifyShapeAttribute(argType) ==
+              tf::AttributeTypeClassifier::Shape::Unsupported) {
+            diagnose(argLoc, "$shape attribute requires TensorShape or "
+                             "TensorShape? but got type '" +
+                             argType->getString() + "'");
+            return nullptr;
+          }
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::TFDataTypeAttribute:
+          if (atc.classifyTFDataTypeAttribute(argType) ==
+              tf::AttributeTypeClassifier::TFDataType::Unsupported) {
+            diagnose(argLoc, "$dtype attribute requires TensorDataType "
+                             "or [TensorDataType] but got type '" +
+                             argType->getString() + "'");
+            return nullptr;
+          }
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::Out:
+          // SILGen generates $out attributes. It does not make sense to specify
+          // them in code.
+          diagnose(argLoc, "$out attributes are not allowed in #tfop's");
+          return nullptr;
+        }
+      }
+
+      // TODO: Check result types.
```

Comment:
Do you mean to address this TODO in a later PR?

Reply:
Yep. I'll add a bug SR-9248